### DISCUSSION TITLE
YALMIP solver

### DISCUSSION THREAD

**Discussion by ArjunM21** [yalmip/YALMIP#1009] [Optimization modelling] [answered]
I am trying to code the 2 methods discussed in the research paper attached. I am trying to code the first method given in the paper. In it, we can see a constraint given in equation 4. I have coded all of it into MATLAB, but I am unable to run it as I get an error message stating that "First argument (F) should be a constraint object". Now to my understanding this is because the constraint itself is not an LMI constraint (it is not linear).  I am quite new to MATLAB so my code is not the best. How do I go about solving this?

The code I have written so far is

`clc;
clear all;
close all;
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0;1 -1 1]; %A matrix from paper

q = sdpvar(8,1);
q0 = [0;1;0;0;1;0;1;0];
Q = diag(q);
Q0 = diag(q0);

Q = double(Q);

G = trace((C/(C'*Q*C))*C');
P = inv((Q*C)'*C);
P0 = inv(((Q-Q0)*C)'*C);
Gdif = C*P*C'*C*P*C';
G0 = trace(C/(C'*Q0*C)*C'); %G(qbar) from the paper
Gdif0 = C*P0*C'*C*P0*C'; %deltaG(q-qbar) from paper

t = G; %same as alpha from the paper
Qt = sum(q);
Objective = t;
Constraints = [G0 + Gdif0 <= t];
% Constraints = [Qt < 4]; %If I run this, I can get a solution 

options=sdpsettings('bnb.solver','gurobi');
sol = solvesdp(Constraints,Objective, options);`

[Optimal sensor placement strategies for large scale.pdf](https://github.com/yalmip/YALMIP/files/7264764/Optimal.sensor.placement.strategies.for.large.scale.pdf)

**Comment by johanlofberg** [comment, marked as answer]
double (which is obsolete and now is called value) cannot be applied before an optimization problem has been solved (or a variable has been manually assigned value) as it is used to extract a solution, which of course isn't available when you call it. Hence Q=double(Q) will simply return NaNs and any further use of Q thereafter will just generate garbage.

It looks like your code is a big confusion between a linearization point q0/Q0 and a decision variable q and associated expression Q. Also (3) is a linear SDP so no reason to linearize. If it is about solving a MISDP, YALMIP can do that for you, and if it is about adding cuts to approximate the MISDP as a MILP, YALMIP can do that too as it has a built-in solver doing that https://yalmip.github.io/The-cutsdp-solver

**Reply by johanlofberg** [reply]
A sensible linearization-based pattern would be

```
q = binvar(8,1)
Q = diag(q)

...
assign(q,q0)
Q0 = value(Q)
Model = sum(q)==M;
while not bored
 create linearized model using q0/Q0/q/Q
 Model = [Model, more cuts from linearized model]
 optimize()
 q0 = value(q)
 Q0 = value(Q)
end
```

**Reply by ArjunM21** [reply]
Thank you for your reply. Some things aren't clear to me. First, your statement regarding (3). The paper discusses some methods to convert it into (4) for solvability, so I am not sure what is wrong. I think cutsdp is what I might be looking for, so I'll go through the article you have shared. Second, for the code that you have given, how do I give constraints? The model part is clear but I always get error while giving constraints. Also, the constraints that I have given in my code does it seem to match the one from paper?

**Reply by johanlofberg** [reply]
The authors in the paper appears to not want to solve the problem in the MISDPform (3), which I question why? It is trivially solved (note that you have defined q as a continuous variable, but it is a binary in the paper). This is trivially solved in<1s using bnb. cutsdp performs worse

```
q = binvar(8,1);
Q = diag(q);
Y = sdpvar(8);
M = 5;
Model = [[Y A;A' A'*Q*A] >= 0,sum(q)==M]
optimize(Model,trace(Y))
```

Instead, they work with the original nonlinear form, and to handle that they linearize it and perform an iteration adding cuts. Your setup is simply not working as you use Q=double(Q) and then use that data (which is Nan, look at Q!) to create the function, when instead you should use the current iterate Q0 which is constructed from q0 (alternative from Q0=value(Q) after q0 has been assigned to q). Then you run a loop to iteratively approximate around last solution, and to get that whole thing working I've shown how to setup the structure. A linear approximation of f(x) is given by dfdx(x0)*(x-x0). You are doing dfdx(NaN)(x-x0) as you are mixing up what Q is (an expression depending on q) and Q0 (a constant constructed from q0, and later by using value() when q and Q has assigned values)

**Reply by johanlofberg** [reply]
and they are not converting (3) into (4), they are converting (2) into (4).

**Reply by ArjunM21** [reply]
Thank you for the reply. So the reason they do this new method is because the size of q and C is for now small, but grows exponentially. This is only an example of this. Supposedly for large scale systems when this is used, the new method is easier and better to solve. 
I have understood my mistake as of now. Although I have not understood how the part of your code works, I rewrote my code from what you suggested. Could you take a look? I am not getting a solution. I am not sure wrt the while loop. As of now, I've tried to implement the condition given in the steps.

`clc;
clear all;
close all;
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0;1 -1 1]; %A matrix from paper

q = binvar(8,1);
Q = diag(q);
q0 = [0;1;0;0;1;0;1;0];
assign(q,q0);
Q0 = value(Q);
G0 = 1;
t = 0;
Model = sum(q)==3;

while (G0 - t < 0.05)
    G = trace((C/(C'*Q*C))*C');
    P = inv((Q*C)'*C);
    P0 = inv(((Q-Q0)*C)'*C);
    Gdif = C*P*C'*C*P*C';
    G0 = trace(C/(C'*Q0*C)*C'); %G(qbar) from the paper
    Gdif0 = C*P0*C'*C*P0*C'; %deltaG(q-qbar) from paper
    t = G; %same as alpha from the paper
    Model = [Model, G0 + Gdif0 <= t];
    ss = optimize();
    q0 = value(q);
    Q0 = value(Q);
    Qt = sum(q0);
end
`

Something seems off for sure.

**Reply by johanlofberg** [reply]
You are still using the decision variable to computedata in the linear cut. The gradient G would be G = trace((C/(C'*Q0*C))*C') etc

You are simply replacing f(x) with f(x0) + df*(x-x0), so the only sensible would be something along the lines of

sdpvar t
Model = sum(q)==3;
while whatever
 gradient = some_function_of(q0,Q0)
 offset= some_function_of(q0,Q0)
 Model = [Model, offset + gradient'*(q-q0) <= t];
 ss = optimize(Model,t)
 q0 = value(q);
  Q0 = value(Q);
end
`

**Reply by ArjunM21** [reply]
Hey thanks for the reply. The code seems to be working but for some reason, I am not getting a solved value. I'll keep trying.

Could you also show me how to code the equation 2 from the paper? As there is no constraints, I am finding a lot of errors in my code.

Thank you.

**Reply by johanlofberg** [reply]
You cannot solve (2) using YALMIP (or any publically available solver), that's the whole reason for deriving the linear equivalent model (3), or using an iterative scheme based on linearizations (4). I have no idea why you say there are no constraints in (2), as there obviously are.

**Reply by ArjunM21** [reply]
Okay that was my mistake, as I meant sdp constraints. How do you think solve eq 2 could be solved then? I'm not looking for a method per say, but what you think.

**Reply by johanlofberg** [reply]
You would simply have to attack it using a general MINLP solver, manually deriving and supply the derivatives w.r.t q for the continuous relaxations solved in the MINLP solver. And there is an implicit SDP constraint in the model in that A'QA has to be positive definite.

For n less than, say, 15, I would not be surprised if enumeration would beat most solver though (at least it would be fast and trivial to implement)., More specifically, for nchoosek(n,M) less than some ten or even hundred thousands, enumeration would probably be fast enough. Ugly though and nothing to brag about...

**Reply by ArjunM21** [reply]
Thank you so much for your help. This has helped me tremendously.
